// Copyright 2024 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

syntax = "proto3";
option optimize_for = LITE_RUNTIME;

import "print_job_options.proto";

// Describes one managed printer configuration.
// A list of `ManagedPrinterConfiguration` messages converted to a JSON array 
// corresponds to the contents `PrintersBulkConfiguration` and `DevicePrinters`
// chromium policies.
// Printers that have a `guid` field which is not in `PrintersBulkAllowlist`
// or `DevicePrintersAllowlist` will be ignored. Invalid printers (e.g.
// missing required fields, invalid URI) will be ignored.
message ManagedPrinterConfiguration {
  // Required. Unique identifier for the printer. This is referenced in the
  // `PrintersBulkConfiguration` and `DevicePrintersAllowlist` policies.
  optional string guid = 1;

  // Required. Printer name visible in the print preview dialog.
  optional string display_name = 2 [json_name = "display_name"];

  // Optional. Printer description visible in the print preview dialog.
  optional string description = 3;

  // Required. The URI used to connect to the printer.
  optional string uri = 4;

  // Defines the source of the PPD file to be used during printer setup.
  message PpdResource {
    oneof resource {
      // The PPD file matching `effective_model` from the PPD index will be
      // used.
      string effective_model = 1 [json_name = "effective_model"];

      // A PPD file will be generated by querying the printer for its
      // capabilities and available options. Suitable for IPP-Everywhere
      // printers.
      bool autoconf = 2;

      // The PPD file hosted at `user_supplied_ppd_uri` will be downloaded and
      // used during printer setup.
      string user_supplied_ppd_uri = 3 [json_name = "user_supplied_ppd_uri"];
    }
  }

  // Required. Source of the PPD file to use during printer setup.
  optional PpdResource ppd_resource = 5 [json_name = "ppd_resource"];

  // Optional. Contains print job options specific to this particular printer.
  optional PrintJobOptions print_job_options = 6
      [json_name = "print_job_options"];
}
